### DISCUSSION TITLE
VKGuide 2 thread

### DISCUSSION THREAD

**Discussion by vblanco20-1** [vblanco20-1/vulkan-guide#92] [Announcements]
VKGuide 2 is now being written. This thread will be a changelog of what happens and the stage of the new version, and a discussion of its features and current prototype code.

Current output of the engine by the end of base tutorial. Scene is taken from godot-tps demo. (0.17 miliseconds of cpu processing time, to draw this scene that has 1700 objects without any caching/fully dynamic rendering. Frame time locked at 144hz for vsync)
![image](https://github.com/vblanco20-1/vulkan-guide/assets/1417000/11533297-cb4c-485f-9c9f-9620dd0d0715)

Vkguide 2 is an almost complete rewrite with the goal to update to vulkan 1.3 and take advantage of its features, alongside some other important updates. Its going to end up being a bit longer than current vkguide, and it has full focus on PC and relatively modern graphics cards.
* No more Renderpasses: Vulkan 1.3 has dynamic rendering as a main feature, so we will no longer do any renderpass code. Instead, we will be dealing with images and drawing to them directly. The code of hello triangle is a fair bit shorter thanks to this.
* Embedding Extra chapters: Descriptor set management and dear imgui ui are going to be part of the tutorial. 
* GLTF Assets: The last chapter of the tutorial is loading real gltfs from blender or other software into the engine.
* Actual useable architecture: Dynamic load and unload of GLTF assets, and completely dynamic high performance rendering.
* Compute shaders from the start: The first rendering we will be done will be through compute shaders. We will not have a hello triangle as first draw, but a hello compute into an image.

In the branch `all-chapters-1.3-wip` you can check the current status of the chapter 0 to 5 codebase and the differences on the engine itself. It contains the code for the first 5 chapters, equivalent to current vkguide, with the addition that it has full scene loading and the possibility of rendering objects dynamically. Unlike the code in the older Engine branch, this one can be run directly, and it includes some example GLTF assets to load. 

After the main 5 chapters are written, an extra 2 chapters will be developed, as an "advanced" extension. Those chapters will continue the tutorial into high performance gpu-driven rendering, and PBR shaders with basic forward lighting. They will be using descriptor indexing and buffer-device-adress.

The chapters will be written with a markdown preprocessor that will directly copy from snippets into the tutorial. This should completely remove the constant issues of example code to explanation desyncs.

Current status:
base code: Alpha stage. Code is done, just still a bit unpolished and will see some minor tweaks.
expansion code: Planning/prototype stage:

Chapter 0: Release candidate
Chapter 1: Release candidate
Chapter 2: Release candidate
Chapter 3: Release candidate
Chapter 4: Release candidate
Chapter 5: Release candidate

Current demo scenes will be amazon lumberyard bistro, godot tps-demo map, and a couple extra maps from [Brickadia](https://brickadia.com/), courtesy of the developers. Even then, more scenes are needed to test the engine fully specially once the draw indirect chapters arrive. If you have some scenes that have a permissive license and would be a great showcase (lots of meshes, lights, lots of materials), then please contact me and show me, that way we can give more "proven to work" demo assets for the project.

**Comment by aer-i** [comment]
Thank you for your work. I needed something like this

**Comment by vblanco20-1** [comment]
Chapters 0-1-2 are now in a early but followable status. They introduce imgui, a double-buffered render loop, and using compute shaders for all drawing.

**Comment by vblanco20-1** [comment]
Chapter 3 is now done. begins drawing a triangle, and loads vertex data through buffer-device-address

**Comment by vblanco20-1** [comment]
Vkguide 2 main timeline is now done, on a beta state. It can be followed now, but there is a lack of polish.

**Comment by vblanco20-1** [comment]
Chapters 0 to 3 now are polished and improved with double-checking. Window resizing has been added there too.

**Comment by vblanco20-1** [comment]
As a christmas present, vkguide 2 is now fully doublechecked with the issues fixed, so the entire thing is now correctly followable with a higher level of polish. This is now a release candidate, and the only thing left is  reviews by other people and some grammar checking on the text.

**Comment by vblanco20-1** [comment]
The website has been rearranged to put the new tutorial front and center, the older version is now under [Legacy Vkguide](https://vkguide.dev/docs/old_vkguide/)

**Comment by LittleGentry** [comment]
Hello my respect! I want to consult some question. Project in engine branch  doesn't work, after cmake, project miss SDL dependency. I saw 
 code about SDL in CMakeList not correct. And, when init_scene(), engine often load .pfb, but there is not .pdf in assets from engine branch, there is only .glts and .bin. Waiting for your response. Thanks.

**Comment by vblanco20-1** [comment]
@LittleGentry  project in engine branch is only there to be a source for the gpu driven chapter. its outdated and missing the assets as i dont have a demo scene to showcase the indirect well on there.

**Comment by WayneTian-42** [comment]
First of all, I sincerely appreciate all your hard work! This tutorial is truly amazing, especially the updated version that incorporates Vulkan 1.3 features to significantly enhance performance. This is not only the biggest difference between your tutorial and others but also its most impressive highlight. Additionally, I’d like to inquire about the plan for the additional chapters. How long will it take for the GPU-Driven chapter to be ready? Is it currently in the coding stage or the text tutorial stage? If possible, could you release the code first? Lastly, thank you again for all your efforts and contributions!
